contents/handbook/engineering/query-performance-optimization.md

```diff
@@ -1,28 +1,82 @@
 ---
-title: ClickHouse query performance optimization
+title: Query performance optimization
 sidebar: Handbook
 showTitle: true
 ---
 
 Making sure PostHog operates fast at scale is key to our success.
 
-This document outlines some tools we have to discover and fix issues.
+This document outlines some best practices to archive good query performance at scale, as well as describing tools and procedures to discover and fix performance issues.
 
-# Tools of the trade
+PostHog uses two different datastores:
 
-## 1. Grafana "Clickhouse queries - by endpoint" dashboard
+- **PostgreSQL**: row-oriented OLTP database, mainly used to store and query datasets using predictable clausole(s). It is likely your best choice if:
+    - the query pattern to access your dataset is predictable
+    - the dataset will likely not grow overtime above (<= 1 TB)
+    - the dataset needs to mutate often (`DELETE`/`UPDATE`)
+    - the query pattern requires joins across multiple tables
 
-https://metrics.posthog.com/d/vo7oCVZ7z/clickhouse-queries-by-endpoint
+- **ClickHouse**: column-oriented OLAP database, used to store large datasets and run on them analytical queries. It is likely your best choice if:
+    - the query pattern to access your dataset is unpredictable
+    - the dataset will likely grow overtime (> 1 TB)
+    - the dataset doesn't need to mutate often (`DELETE`/`UPDATE`)
+    - the query pattern doesn't requires joins across multiple tables
 
-This dashboard gives a breakdown of how things are looking reliability and performance-wise.
+## PostgreSQL
 
+#### Coding best practices
+
+1. only ask for the field(s) you need: `SELECT name, surname` is better than `SELECT *` (the latter is only helpful in few edge cases)
+
+1. only ask for the row(s) you need: use a `LIMIT` condition at the end of your query
+
+1. (if possible) avoid explicit transactions: if you can't, keep them small since transactions lock the processing tables data and may result in deadlocks (super discouraged to use them in application hot paths)
+
+1. (if possible) avoid `JOIN`
+
+1. **always avoid** foreign keys: they are in your way to shard your database, your app is accustomed to rely on them to maintain data integrity instead of doing it on its own, they have performance impact (they add an additional lookup for each insert/delete), they lead to unpredictable performance, ...
```

Comment:
This is tricky with the Django ORM. If we go down this route might we need to consider just writing sql?

Reply:
I know, we might end up (in the future) to write our own classes to access the data layer

Reply:
Could we make the doc be suggestions of what we should do now. Currently we seem to be using foreign keys & if I shouldn't use it, what should I do instead?

Reply:
I think the goal should be to don't add more FK but still keep the one we have (and maybe in the future we will work on removing them)